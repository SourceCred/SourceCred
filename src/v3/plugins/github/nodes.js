// @flow

import {NodeAddress, type NodeAddressT} from "../../core/graph";

export opaque type RawAddress: NodeAddressT = NodeAddressT;

const GITHUB_PREFIX = NodeAddress.fromParts(["sourcecred", "github"]);
function githubAddress(...parts: string[]): RawAddress {
  return NodeAddress.append(GITHUB_PREFIX, ...parts);
}

export type RepoAddress = {|
  +type: "REPO",
  +owner: string,
  +name: string,
|};
export type IssueAddress = {|
  +type: "ISSUE",
  +repo: RepoAddress,
  +number: string,
|};
export type PullAddress = {|
  +type: "PULL",
  +repo: RepoAddress,
  +number: string,
|};
export type ReviewAddress = {|
  +type: "REVIEW",
  +pull: PullAddress,
  +fragment: string,
|};
export type CommentAddress = {|
  +type: "COMMENT",
  +parent: IssueAddress | PullAddress | ReviewAddress,
  +fragment: string,
|};
export type UserlikeAddress = {|
  +type: "USERLIKE",
  +login: string,
|};

export type StructuredAddress =
  | RepoAddress
  | IssueAddress
  | PullAddress
  | ReviewAddress
  | CommentAddress
  | UserlikeAddress;

// Each of these types has 0 or more "AUTHORS" edges, each of which
// leads to a UserlikeAddress.  Note: It is not true that every
// Authorable has at least one author, as when GitHub accounts are
// deleted, they leave behind posts without authors.
export type AuthorableAddress =
  | IssueAddress
  | PullAddress
  | ReviewAddress
  | CommentAddress;

// Each of these types has text content, which means
// it may be the source of a reference to a ReferentAddress.
export type TextContentAddress =
  | IssueAddress
  | PullAddress
  | ReviewAddress
  | CommentAddress;

// Each of these types may be referred to by something
// with text content.
export type ReferentAddress =
  | RepoAddress
  | IssueAddress
  | PullAddress
  | ReviewAddress
  | CommentAddress
  | UserlikeAddress;

// Each of these types is structurally the child of some
// entity with a ParentAddress.
export type ChildAddress =
  | IssueAddress // child of RepoAddress
  | PullAddress // child of RepoAddress
  | ReviewAddress // child of PullAddress
  | CommentAddress; // child of IssueAddress, PullAddress, or ReviewAddress

// Each of these types may have child nodes (see ChildAddress).
export type ParentAddress =
  | RepoAddress
  | IssueAddress
  | PullAddress
  | ReviewAddress;

export function fromRaw(x: RawAddress): StructuredAddress {
  function fail() {
    return new Error(`Bad address: ${NodeAddress.toString(x)}`);
  }
  if (!NodeAddress.hasPrefix(x, GITHUB_PREFIX)) {
    throw fail();
  }
  const [_unused_sc, _unused_gh, kind, ...rest] = NodeAddress.toParts(x);
  switch (kind) {
    case "repo": {
      if (rest.length !== 2) {
        throw fail();
      }
      const [owner, name] = rest;
      return {type: "REPO", owner, name};
    }
    case "issue": {
      if (rest.length !== 3) {
        throw fail();
      }
      const [owner, name, number] = rest;
      const repo = {type: "REPO", owner, name};
      return {type: "ISSUE", repo, number};
    }
    case "pull": {
      if (rest.length !== 3) {
        throw fail();
      }
      const [owner, name, number] = rest;
      const repo = {type: "REPO", owner, name};
      return {type: "PULL", repo, number};
    }
    case "review": {
      if (rest.length !== 4) {
        throw fail();
      }
      const [owner, name, pullNumber, fragment] = rest;
      const repo = {type: "REPO", owner, name};
      const pull = {type: "PULL", repo, number: pullNumber};
      return {type: "REVIEW", pull, fragment};
    }
    case "comment": {
      if (rest.length < 1) {
        throw fail();
      }
      const [subkind, ...subrest] = rest;
      switch (subkind) {
        case "issue": {
          if (subrest.length !== 4) {
            throw fail();
          }
          const [owner, name, issueNumber, fragment] = subrest;
          const repo = {type: "REPO", owner, name};
          const issue = {type: "ISSUE", repo, number: issueNumber};
          return {type: "COMMENT", parent: issue, fragment};
        }
        case "pull": {
          if (subrest.length !== 4) {
            throw fail();
          }
          const [owner, name, pullNumber, fragment] = subrest;
          const repo = {type: "REPO", owner, name};
          const pull = {type: "PULL", repo, number: pullNumber};
          return {type: "COMMENT", parent: pull, fragment};
        }
        case "review": {
          if (subrest.length !== 5) {
            throw fail();
          }
          const [owner, name, pullNumber, reviewFragment, fragment] = subrest;
          const repo = {type: "REPO", owner, name};
          const pull = {type: "PULL", repo, number: pullNumber};
          const review = {type: "REVIEW", pull, fragment: reviewFragment};
          return {type: "COMMENT", parent: review, fragment};
        }
        default:
          throw fail();
      }
    }
    case "userlike": {
      if (rest.length !== 1) {
        throw fail();
      }
      const [login] = rest;
      return {type: "USERLIKE", login};
    }
    default:
      throw fail();
  }
}

export function toRaw(x: StructuredAddress): RawAddress {
  switch (x.type) {
    case "REPO":
      return githubAddress("repo", x.owner, x.name);
    case "ISSUE":
      return githubAddress("issue", x.repo.owner, x.repo.name, x.number);
    case "PULL":
      return githubAddress("pull", x.repo.owner, x.repo.name, x.number);
    case "REVIEW":
      return githubAddress(
        "review",
        x.pull.repo.owner,
        x.pull.repo.name,
        x.pull.number,
        x.fragment
      );
    case "COMMENT":
      switch (x.parent.type) {
        case "ISSUE":
          return githubAddress(
            "comment",
            "issue",
            x.parent.repo.owner,
            x.parent.repo.name,
            x.parent.number,
            x.fragment
          );
        case "PULL":
          return githubAddress(
            "comment",
            "pull",
            x.parent.repo.owner,
            x.parent.repo.name,
            x.parent.number,
            x.fragment
          );
        case "REVIEW":
          return githubAddress(
            "comment",
            "review",
            x.parent.pull.repo.owner,
            x.parent.pull.repo.name,
            x.parent.pull.number,
            x.parent.fragment,
            x.fragment
          );
        default:
          // eslint-disable-next-line no-unused-expressions
          (x.parent.type: empty);
          throw new Error(`Bad comment parent type: ${x.parent.type}`);
      }
    case "USERLIKE":
      return githubAddress("userlike", x.login);
    default:
      // eslint-disable-next-line no-unused-expressions
      (x.type: empty);
      throw new Error(`Unexpected type ${x.type}`);
  }
}
