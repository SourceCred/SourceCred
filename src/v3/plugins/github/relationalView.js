// @flow

import * as N from "./nodes";
import * as Q from "./graphql";
import * as GitNode from "../git/nodes";

import {
  reviewUrlToId,
  issueCommentUrlToId,
  pullCommentUrlToId,
  reviewCommentUrlToId,
} from "./urlIdParse";

export type RepoEntry = {|
  +type: typeof N.REPO_TYPE,
  +address: N.RepoAddress,
  +url: string,
  +issues: IssueEntry[],
  +pulls: PullEntry[],
|};

export type IssueEntry = {|
  +type: typeof N.ISSUE_TYPE,
  +address: N.IssueAddress,
  +title: string,
  +body: string,
  +url: string,
  +comments: CommentEntry[],
  +nominalAuthor: ?UserlikeEntry,
|};

export type PullEntry = {|
  +type: typeof N.PULL_TYPE,
  +address: N.PullAddress,
  +title: string,
  +body: string,
  +url: string,
  +comments: CommentEntry[],
  +reviews: ReviewEntry[],
  +mergedAs: ?GitNode.CommitAddress,
  +nominalAuthor: ?UserlikeEntry,
|};

export type ReviewEntry = {|
  +type: typeof N.REVIEW_TYPE,
  +address: N.ReviewAddress,
  +body: string,
  +url: string,
  +comments: CommentEntry[],
  +state: Q.ReviewState,
  +nominalAuthor: ?UserlikeEntry,
|};

export type CommentEntry = {|
  +type: typeof N.COMMENT_TYPE,
  +address: N.CommentAddress,
  +body: string,
  +url: string,
  +nominalAuthor: ?UserlikeEntry,
|};

export type UserlikeEntry = {|
  +type: typeof N.USERLIKE_TYPE,
  +address: N.UserlikeAddress,
  +url: string,
|};

export type Entry =
  | RepoEntry
  | IssueEntry
  | PullEntry
  | ReviewEntry
  | CommentEntry
  | UserlikeEntry;

export class RelationalView {
  _repos: Map<N.RawAddress, RepoEntry>;
  _issues: Map<N.RawAddress, IssueEntry>;
  _pulls: Map<N.RawAddress, PullEntry>;
  _comments: Map<N.RawAddress, CommentEntry>;
  _reviews: Map<N.RawAddress, ReviewEntry>;
  _userlikes: Map<N.RawAddress, UserlikeEntry>;

  constructor(data: Q.GithubResponseJSON) {
    this._repos = new Map();
    this._issues = new Map();
    this._pulls = new Map();
    this._comments = new Map();
    this._reviews = new Map();
    this._userlikes = new Map();
    this._addRepo(data.repository);
  }

  repos(): Iterator<RepoEntry> {
    return this._repos.values();
  }
  repo(address: N.RepoAddress): ?RepoEntry {
    return this._repos.get(N.toRaw(address));
  }

  issues(): Iterator<IssueEntry> {
    return this._issues.values();
  }
  issue(address: N.IssueAddress): ?IssueEntry {
    return this._issues.get(N.toRaw(address));
  }

  pulls(): Iterator<PullEntry> {
    return this._pulls.values();
  }
  pull(address: N.PullAddress): ?PullEntry {
    return this._pulls.get(N.toRaw(address));
  }

  comments(): Iterator<CommentEntry> {
    return this._comments.values();
  }
  comment(address: N.CommentAddress): ?CommentEntry {
    return this._comments.get(N.toRaw(address));
  }

  reviews(): Iterator<ReviewEntry> {
    return this._reviews.values();
  }
  review(address: N.ReviewAddress): ?ReviewEntry {
    return this._reviews.get(N.toRaw(address));
  }

  userlikes(): Iterator<UserlikeEntry> {
    return this._userlikes.values();
  }
  userlike(address: N.UserlikeAddress): ?UserlikeEntry {
    return this._userlikes.get(N.toRaw(address));
  }

  _addRepo(json: Q.RepositoryJSON) {
    const address: N.RepoAddress = {
      type: N.REPO_TYPE,
      owner: json.owner.login,
      name: json.name,
    };
    const entry: RepoEntry = {
      address,
      url: json.url,
      issues: json.issues.nodes.map((x) => this._addIssue(address, x)),
      pulls: json.pulls.nodes.map((x) => this._addPull(address, x)),
      type: N.REPO_TYPE,
    };
    const raw = N.toRaw(address);
    this._repos.set(raw, entry);
  }

  _addIssue(repo: N.RepoAddress, json: Q.IssueJSON): IssueEntry {
    const address: N.IssueAddress = {
      type: N.ISSUE_TYPE,
      number: String(json.number),
      repo,
    };
    const entry: IssueEntry = {
      address,
      url: json.url,
      comments: json.comments.nodes.map((x) => this._addComment(address, x)),
      nominalAuthor: this._addNullableAuthor(json.author),
      body: json.body,
      title: json.title,
      type: N.ISSUE_TYPE,
    };
    this._issues.set(N.toRaw(address), entry);
    return entry;
  }

  _addPull(repo: N.RepoAddress, json: Q.PullJSON): PullEntry {
    const address: N.PullAddress = {
      type: N.PULL_TYPE,
      number: String(json.number),
      repo,
    };
    const mergedAs =
      json.mergeCommit == null
        ? null
        : {
            type: GitNode.COMMIT_TYPE,
            hash: json.mergeCommit.oid,
          };

    const entry: PullEntry = {
      address,
      url: json.url,
      comments: json.comments.nodes.map((x) => this._addComment(address, x)),
      reviews: json.reviews.nodes.map((x) => this._addReview(address, x)),
      nominalAuthor: this._addNullableAuthor(json.author),
      body: json.body,
      title: json.title,
      mergedAs,
      type: N.PULL_TYPE,
    };
    this._pulls.set(N.toRaw(address), entry);
    return entry;
  }

  _addReview(pull: N.PullAddress, json: Q.ReviewJSON): ReviewEntry {
    const address: N.ReviewAddress = {
      type: N.REVIEW_TYPE,
      id: reviewUrlToId(json.url),
      pull,
    };
    const entry: ReviewEntry = {
      address,
      url: json.url,
      state: json.state,
      comments: json.comments.nodes.map((x) => this._addComment(address, x)),
      body: json.body,
      nominalAuthor: this._addNullableAuthor(json.author),
      type: N.REVIEW_TYPE,
    };
    this._reviews.set(N.toRaw(address), entry);
    return entry;
  }

  _addComment(
    parent: N.IssueAddress | N.PullAddress | N.ReviewAddress,
    json: Q.CommentJSON
  ): CommentEntry {
    const id = (function() {
      switch (parent.type) {
        case N.ISSUE_TYPE:
          return issueCommentUrlToId(json.url);
        case N.PULL_TYPE:
          return pullCommentUrlToId(json.url);
        case N.REVIEW_TYPE:
          return reviewCommentUrlToId(json.url);
        default:
          // eslint-disable-next-line no-unused-expressions
          (parent.type: empty);
          throw new Error(`Unexpected comment parent type: ${parent.type}`);
      }
    })();
    const address: N.CommentAddress = {type: N.COMMENT_TYPE, id, parent};
    const entry: CommentEntry = {
      address,
      url: json.url,
      nominalAuthor: this._addNullableAuthor(json.author),
      body: json.body,
      type: N.COMMENT_TYPE,
    };
    this._comments.set(N.toRaw(address), entry);
    return entry;
  }

  _addNullableAuthor(json: Q.NullableAuthorJSON): ?UserlikeEntry {
    if (json == null) {
      return null;
    } else {
      const address: N.UserlikeAddress = {
        type: N.USERLIKE_TYPE,
        login: json.login,
      };
      const entry: UserlikeEntry = {
        address,
        url: json.url,
        type: N.USERLIKE_TYPE,
      };
      this._userlikes.set(N.toRaw(address), entry);
      return entry;
    }
  }
}
