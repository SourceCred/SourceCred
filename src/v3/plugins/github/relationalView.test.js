// @flow

import * as R from "./relationalView";
import * as N from "./nodes";

describe("plugins/github/relationalView", () => {
  const data = require("./example/example-github");
  // Sharing this state is OK because it's just a view - no mutation allowed!
  const view = new R.RelationalView(data);

  function hasEntities(name, method) {
    describe(name, () => {
      const all = Array.from(method());
      it(`has expected number of ${name}`, () => {
        expect(all.length).toMatchSnapshot();
      });
      it("have expected urls", () => {
        expect(all.map((x) => x.url())).toMatchSnapshot();
      });
    });
  }

  function has(name, method) {
    it(`has ${name}`, () => {
      const element = method();
      let snapshot;
      if (element instanceof R.Entity) {
        // element is an Entity. Entities have pointers to the RelationalView,
        // and it would pollute our snapshot horribly. Just show the url.
        snapshot = {url: element.url()};
      } else {
        snapshot = element;
      }
      expect(snapshot).toMatchSnapshot();
    });
  }

  describe("RelationalView", () => {
    function hasEntityMethods<T: R.Entity<any>>(
      name,
      getAll: () => Iterator<T>,
      get: (x: $Call<$PropertyType<T, "address">>) => ?T
    ) {
      describe(`entity: ${name}`, () => {
        const all = Array.from(getAll());
        it("has expected number of them", () => {
          expect(all.length).not.toEqual(0);
          expect(all.length).toMatchSnapshot();
        });
        const one = all[0];
        it("they are retrievable by address", () => {
          expect(get(one.address())).toEqual(one);
        });
        it("they have expected urls", () => {
          expect(all.map((x) => x.url())).toMatchSnapshot();
        });
      });
    }
    hasEntityMethods("repos", () => view.repos(), (x) => view.repo(x));
    hasEntityMethods("issues", () => view.issues(), (x) => view.issue(x));
    hasEntityMethods("pulls", () => view.pulls(), (x) => view.pull(x));
    hasEntityMethods("reviews", () => view.reviews(), (x) => view.review(x));
    hasEntityMethods("comments", () => view.comments(), (x) => view.comment(x));
    hasEntityMethods(
      "userlikes",
      () => view.userlikes(),
      (x) => view.userlike(x)
    );
  });

  const repo = view.repo({
    type: N.REPO_TYPE,
    owner: "sourcecred",
    name: "example-github",
  });
  if (repo == null) {
    throw new Error("Error: sourcecred/example-github must exist!");
  }
  describe("Repo", () => {
    const entity = repo;
    has("owner", () => entity.owner());
    has("name", () => entity.name());
    has("url", () => entity.url());
    hasEntities("issues", () => entity.issues());
    hasEntities("pulls", () => entity.pulls());
  });

  const issue = Array.from(repo.issues())[1];
  describe("Issue", () => {
    const entity = issue;
    has("number", () => entity.number());
    has("body", () => entity.body());
    has("title", () => entity.title());
    has("url", () => entity.url());
    has("parent", () => entity.parent());
    hasEntities("comments", () => entity.comments());
    hasEntities("authors", () => entity.authors());
  });

  const pull = Array.from(repo.pulls())[1];
  describe("Pull", () => {
    const entity = pull;
    has("number", () => entity.number());
    has("body", () => entity.body());
    has("title", () => entity.title());
    has("url", () => entity.url());
    has("parent", () => entity.parent());
    has("mergedAs", () => entity.mergedAs());
    hasEntities("reviews", () => entity.reviews());
    hasEntities("comments", () => entity.comments());
    hasEntities("authors", () => entity.authors());
  });

  const review = Array.from(pull.reviews())[0];
  describe("Review", () => {
    const entity = review;
    has("body", () => entity.body());
    has("url", () => entity.url());
    has("state", () => entity.state());
    has("parent", () => entity.parent());
    hasEntities("comments", () => entity.comments());
    hasEntities("authors", () => entity.authors());
  });

  const comment = Array.from(review.comments())[0];
  describe("Comment", () => {
    const entity = comment;
    has("body", () => entity.body());
    has("url", () => entity.url());
    has("parent", () => entity.parent());
    hasEntities("authors", () => entity.authors());
  });

  const userlike = Array.from(review.authors())[0];
  describe("Userlike", () => {
    const entity = userlike;
    has("login", () => entity.login());
    has("url", () => entity.url());
  });

  describe("comment parent differentiation", () => {
    function hasCorrectParent(name, parent) {
      it(name, () => {
        const comment = Array.from(parent.comments())[0];
        expect(comment).toEqual(expect.anything());
        const actualParent = comment.parent();
        expect(parent.address()).toEqual(actualParent.address());
      });
    }
    hasCorrectParent("issue", issue);
    hasCorrectParent("pull", pull);
    hasCorrectParent("review", review);
  });

  describe("reference detection", () => {
    // create url->url reference maps, for convenient snapshot readability
    const allReferences: Map<string, Set<string>> = new Map();
    let nReferences = 0;
    for (const referrer of view.textContentEntities()) {
      const references = new Set();
      allReferences.set(referrer.url(), references);
      for (const referenced of referrer.references()) {
        references.add(referenced.url());
        nReferences++;
      }
    }

    it("references match snapshot", () => {
      const everyReference = [];
      for (const [referrerUrl, referencedUrls] of allReferences) {
        for (const referencedUrl of referencedUrls) {
          everyReference.push({from: referrerUrl, to: referencedUrl});
        }
      }
      expect(everyReference).toMatchSnapshot();
    });

    it("correspondence between references() and referencedBy()", () => {
      let nFoundReferences = 0;
      for (const referent of view.referentEntities()) {
        for (const referrer of referent.referencedBy()) {
          nFoundReferences++;
          const srcUrl = referrer.url();
          const dstUrl = referent.url();
          const actualRefsFromSrc = allReferences.get(srcUrl);
          if (actualRefsFromSrc == null) {
            throw new Error(`Expected refs for ${srcUrl}`);
          }
          if (!actualRefsFromSrc.has(dstUrl)) {
            throw new Error(`Expected ref from ${srcUrl} to ${dstUrl}`);
          }
        }
      }
      expect(nFoundReferences).toEqual(nReferences);
    });
  });
});
