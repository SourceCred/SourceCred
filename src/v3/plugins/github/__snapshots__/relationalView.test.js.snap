// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins/github/relationalView Comment authors has expected number of authors 1`] = `1`;

exports[`plugins/github/relationalView Comment authors have expected urls 1`] = `
Array [
  "https://github.com/wchargin",
]
`;

exports[`plugins/github/relationalView Comment has body 1`] = `"seems a bit capricious"`;

exports[`plugins/github/relationalView Comment has parent 1`] = `
Object {
  "url": "https://github.com/sourcecred/example-github/pull/5#pullrequestreview-100313899",
}
`;

exports[`plugins/github/relationalView Comment has url 1`] = `"https://github.com/sourcecred/example-github/pull/5#discussion_r171460198"`;

exports[`plugins/github/relationalView Issue authors has expected number of authors 1`] = `1`;

exports[`plugins/github/relationalView Issue authors have expected urls 1`] = `
Array [
  "https://github.com/decentralion",
]
`;

exports[`plugins/github/relationalView Issue comments has expected number of comments 1`] = `8`;

exports[`plugins/github/relationalView Issue comments have expected urls 1`] = `
Array [
  "https://github.com/sourcecred/example-github/issues/2#issuecomment-373768703",
  "https://github.com/sourcecred/example-github/issues/2#issuecomment-373768850",
  "https://github.com/sourcecred/example-github/issues/2#issuecomment-385576185",
  "https://github.com/sourcecred/example-github/issues/2#issuecomment-385576220",
  "https://github.com/sourcecred/example-github/issues/2#issuecomment-385576248",
  "https://github.com/sourcecred/example-github/issues/2#issuecomment-385576273",
  "https://github.com/sourcecred/example-github/issues/2#issuecomment-385576920",
  "https://github.com/sourcecred/example-github/issues/2#issuecomment-385576936",
]
`;

exports[`plugins/github/relationalView Issue has body 1`] = `"This issue references another issue, namely #1"`;

exports[`plugins/github/relationalView Issue has number 1`] = `"2"`;

exports[`plugins/github/relationalView Issue has parent 1`] = `
Object {
  "url": "https://github.com/sourcecred/example-github",
}
`;

exports[`plugins/github/relationalView Issue has title 1`] = `"A referencing issue."`;

exports[`plugins/github/relationalView Issue has url 1`] = `"https://github.com/sourcecred/example-github/issues/2"`;

exports[`plugins/github/relationalView Pull authors has expected number of authors 1`] = `1`;

exports[`plugins/github/relationalView Pull authors have expected urls 1`] = `
Array [
  "https://github.com/decentralion",
]
`;

exports[`plugins/github/relationalView Pull comments has expected number of comments 1`] = `1`;

exports[`plugins/github/relationalView Pull comments have expected urls 1`] = `
Array [
  "https://github.com/sourcecred/example-github/pull/5#issuecomment-396430464",
]
`;

exports[`plugins/github/relationalView Pull has body 1`] = `
"@wchargin could you please do the following:
- add a commit comment
- add a review comment requesting some trivial change
- i'll change it
- then approve the pr"
`;

exports[`plugins/github/relationalView Pull has mergedAs 1`] = `
Object {
  "hash": "6d5b3aa31ebb68a06ceb46bbd6cf49b6ccd6f5e6",
  "type": "COMMIT",
}
`;

exports[`plugins/github/relationalView Pull has number 1`] = `"5"`;

exports[`plugins/github/relationalView Pull has parent 1`] = `
Object {
  "url": "https://github.com/sourcecred/example-github",
}
`;

exports[`plugins/github/relationalView Pull has title 1`] = `"This pull request will be more contentious. I can feel it..."`;

exports[`plugins/github/relationalView Pull has url 1`] = `"https://github.com/sourcecred/example-github/pull/5"`;

exports[`plugins/github/relationalView Pull reviews has expected number of reviews 1`] = `2`;

exports[`plugins/github/relationalView Pull reviews have expected urls 1`] = `
Array [
  "https://github.com/sourcecred/example-github/pull/5#pullrequestreview-100313899",
  "https://github.com/sourcecred/example-github/pull/5#pullrequestreview-100314038",
]
`;

exports[`plugins/github/relationalView RelationalView entity: comments has expected number of them 1`] = `14`;

exports[`plugins/github/relationalView RelationalView entity: comments they have expected urls 1`] = `
Array [
  "https://github.com/sourcecred/example-github/issues/2#issuecomment-373768703",
  "https://github.com/sourcecred/example-github/issues/2#issuecomment-373768850",
  "https://github.com/sourcecred/example-github/issues/2#issuecomment-385576185",
  "https://github.com/sourcecred/example-github/issues/2#issuecomment-385576220",
  "https://github.com/sourcecred/example-github/issues/2#issuecomment-385576248",
  "https://github.com/sourcecred/example-github/issues/2#issuecomment-385576273",
  "https://github.com/sourcecred/example-github/issues/2#issuecomment-385576920",
  "https://github.com/sourcecred/example-github/issues/2#issuecomment-385576936",
  "https://github.com/sourcecred/example-github/issues/6#issuecomment-373768442",
  "https://github.com/sourcecred/example-github/issues/6#issuecomment-373768538",
  "https://github.com/sourcecred/example-github/issues/6#issuecomment-385223316",
  "https://github.com/sourcecred/example-github/pull/3#issuecomment-369162222",
  "https://github.com/sourcecred/example-github/pull/5#issuecomment-396430464",
  "https://github.com/sourcecred/example-github/pull/5#discussion_r171460198",
]
`;

exports[`plugins/github/relationalView RelationalView entity: issues has expected number of them 1`] = `6`;

exports[`plugins/github/relationalView RelationalView entity: issues they have expected urls 1`] = `
Array [
  "https://github.com/sourcecred/example-github/issues/1",
  "https://github.com/sourcecred/example-github/issues/2",
  "https://github.com/sourcecred/example-github/issues/4",
  "https://github.com/sourcecred/example-github/issues/6",
  "https://github.com/sourcecred/example-github/issues/7",
  "https://github.com/sourcecred/example-github/issues/8",
]
`;

exports[`plugins/github/relationalView RelationalView entity: pulls has expected number of them 1`] = `3`;

exports[`plugins/github/relationalView RelationalView entity: pulls they have expected urls 1`] = `
Array [
  "https://github.com/sourcecred/example-github/pull/3",
  "https://github.com/sourcecred/example-github/pull/5",
  "https://github.com/sourcecred/example-github/pull/9",
]
`;

exports[`plugins/github/relationalView RelationalView entity: repos has expected number of them 1`] = `1`;

exports[`plugins/github/relationalView RelationalView entity: repos they have expected urls 1`] = `
Array [
  "https://github.com/sourcecred/example-github",
]
`;

exports[`plugins/github/relationalView RelationalView entity: reviews has expected number of them 1`] = `2`;

exports[`plugins/github/relationalView RelationalView entity: reviews they have expected urls 1`] = `
Array [
  "https://github.com/sourcecred/example-github/pull/5#pullrequestreview-100313899",
  "https://github.com/sourcecred/example-github/pull/5#pullrequestreview-100314038",
]
`;

exports[`plugins/github/relationalView RelationalView entity: userlikes has expected number of them 1`] = `2`;

exports[`plugins/github/relationalView RelationalView entity: userlikes they have expected urls 1`] = `
Array [
  "https://github.com/decentralion",
  "https://github.com/wchargin",
]
`;

exports[`plugins/github/relationalView Repo has name 1`] = `"example-github"`;

exports[`plugins/github/relationalView Repo has owner 1`] = `"sourcecred"`;

exports[`plugins/github/relationalView Repo has url 1`] = `"https://github.com/sourcecred/example-github"`;

exports[`plugins/github/relationalView Repo issues has expected number of issues 1`] = `6`;

exports[`plugins/github/relationalView Repo issues have expected urls 1`] = `
Array [
  "https://github.com/sourcecred/example-github/issues/1",
  "https://github.com/sourcecred/example-github/issues/2",
  "https://github.com/sourcecred/example-github/issues/4",
  "https://github.com/sourcecred/example-github/issues/6",
  "https://github.com/sourcecred/example-github/issues/7",
  "https://github.com/sourcecred/example-github/issues/8",
]
`;

exports[`plugins/github/relationalView Repo pulls has expected number of pulls 1`] = `3`;

exports[`plugins/github/relationalView Repo pulls have expected urls 1`] = `
Array [
  "https://github.com/sourcecred/example-github/pull/3",
  "https://github.com/sourcecred/example-github/pull/5",
  "https://github.com/sourcecred/example-github/pull/9",
]
`;

exports[`plugins/github/relationalView Review authors has expected number of authors 1`] = `1`;

exports[`plugins/github/relationalView Review authors have expected urls 1`] = `
Array [
  "https://github.com/wchargin",
]
`;

exports[`plugins/github/relationalView Review comments has expected number of comments 1`] = `1`;

exports[`plugins/github/relationalView Review comments have expected urls 1`] = `
Array [
  "https://github.com/sourcecred/example-github/pull/5#discussion_r171460198",
]
`;

exports[`plugins/github/relationalView Review has body 1`] = `"hmmm.jpg"`;

exports[`plugins/github/relationalView Review has parent 1`] = `
Object {
  "url": "https://github.com/sourcecred/example-github/pull/5",
}
`;

exports[`plugins/github/relationalView Review has state 1`] = `"CHANGES_REQUESTED"`;

exports[`plugins/github/relationalView Review has url 1`] = `"https://github.com/sourcecred/example-github/pull/5#pullrequestreview-100313899"`;

exports[`plugins/github/relationalView Userlike has login 1`] = `"wchargin"`;

exports[`plugins/github/relationalView Userlike has url 1`] = `"https://github.com/wchargin"`;

exports[`plugins/github/relationalView reference detection references match snapshot 1`] = `
Array [
  Object {
    "from": "https://github.com/sourcecred/example-github/issues/2",
    "to": "https://github.com/sourcecred/example-github/issues/1",
  },
  Object {
    "from": "https://github.com/sourcecred/example-github/pull/5",
    "to": "https://github.com/wchargin",
  },
  Object {
    "from": "https://github.com/sourcecred/example-github/pull/9",
    "to": "https://github.com/wchargin",
  },
  Object {
    "from": "https://github.com/sourcecred/example-github/issues/2#issuecomment-373768703",
    "to": "https://github.com/sourcecred/example-github/issues/6",
  },
  Object {
    "from": "https://github.com/sourcecred/example-github/issues/2#issuecomment-373768850",
    "to": "https://github.com/sourcecred/example-github/issues/6#issuecomment-373768538",
  },
  Object {
    "from": "https://github.com/sourcecred/example-github/issues/2#issuecomment-385576185",
    "to": "https://github.com/sourcecred/example-github/pull/5",
  },
  Object {
    "from": "https://github.com/sourcecred/example-github/issues/2#issuecomment-385576220",
    "to": "https://github.com/sourcecred/example-github/pull/5#pullrequestreview-100313899",
  },
  Object {
    "from": "https://github.com/sourcecred/example-github/issues/2#issuecomment-385576248",
    "to": "https://github.com/sourcecred/example-github/pull/5#discussion_r171460198",
  },
  Object {
    "from": "https://github.com/sourcecred/example-github/issues/2#issuecomment-385576273",
    "to": "https://github.com/wchargin",
  },
  Object {
    "from": "https://github.com/sourcecred/example-github/issues/2#issuecomment-385576920",
    "to": "https://github.com/sourcecred/example-github/issues/1",
  },
  Object {
    "from": "https://github.com/sourcecred/example-github/issues/2#issuecomment-385576920",
    "to": "https://github.com/sourcecred/example-github/issues/2",
  },
  Object {
    "from": "https://github.com/sourcecred/example-github/issues/2#issuecomment-385576920",
    "to": "https://github.com/sourcecred/example-github/pull/3",
  },
  Object {
    "from": "https://github.com/sourcecred/example-github/issues/2#issuecomment-385576920",
    "to": "https://github.com/sourcecred/example-github/pull/5#discussion_r171460198",
  },
  Object {
    "from": "https://github.com/sourcecred/example-github/issues/2#issuecomment-385576920",
    "to": "https://github.com/sourcecred/example-github/pull/5#pullrequestreview-100313899",
  },
  Object {
    "from": "https://github.com/sourcecred/example-github/issues/6#issuecomment-385223316",
    "to": "https://github.com/sourcecred/example-github/issues/2",
  },
  Object {
    "from": "https://github.com/sourcecred/example-github/pull/3#issuecomment-369162222",
    "to": "https://github.com/sourcecred/example-github/issues/2",
  },
]
`;
